lib/internal/assert/snapshot.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+
+const {
+  ArrayPrototypeMap,
+  Boolean,
+  SafeMap,
+  SafeSet,
+  StringPrototypeIndexOf,
+  StringPrototypeSlice,
+  StringPrototypeSplit,
+  StringPrototypeReplace,
+  Symbol,
+} = primordials;
+
+const { codes: { ERR_ASSERT_SNAPSHOT_NOT_SUPPORTED } } = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const { inspect } = require('internal/util/inspect');
+const { validateString } = require('internal/validators');
+const { once } = require('events');
+const { createReadStream, createWriteStream } = require('fs');
+const path = require('path');
+const assert = require('assert');
+
+const kUpdateSnapshot = Boolean(process.env.NODE_UPDATE_SNAPSHOT);
+const kInitialSnapshot = Symbol('kInitialSnapshot');
+const kDefaultDelimiter = '\n#*#*#*$#*#*#*#*#*#*$#*#*#\n';
+
+function defaultSnapshotPath() {
+  if (process.mainModule) {
+    const { dir, name } = path.parse(process.mainModule.filename);
+    return path.join(dir, `${name}.snapshot`);
+  }
+  if (!process.argv[1]) {
+    throw new ERR_ASSERT_SNAPSHOT_NOT_SUPPORTED();
+  }
+  const { dir, name } = path.parse(process.argv[1]);
+  return path.join(dir, `${name}.snapshot`);
+}
+
+let _source;
+function lazySource() {
+  _source = _source ?? createReadStream(defaultSnapshotPath(), { encoding: 'utf8' });
+  return _source;
+}
+
+let _target;
+function lazyTarget() {
+  _target = _target ?? createWriteStream(defaultSnapshotPath(), { encoding: 'utf8' });
+  return _target;
+}
```

Comment:
How does that work? A stream cannot be used twice AFAIK.

Reply:
it is only created once and used once - but I don't want the stream to be created if it is not needed.
I will clean up `_target` as it was from a previous implementation where it was configurabel

Reply:
what do you mean cannot be used twice?
the stream is created once only since `_target = _target ?? ...` prevents it from creating multiple times,
and writing to it CAN happen multiple times, and it does every time you call `assert.snapshot()`

Reply:
```suggestion
let _source;
function lazySource() {
  _source ??= createReadStream(defaultSnapshotPath(), { encoding: 'utf8' });
  return _source;
}

let _target;
function lazyTarget() {
  _target ??= createWriteStream(defaultSnapshotPath(), { encoding: 'utf8' });
  return _target;
}
```

Reply:
OK for the write stream, it really bugs me that we can potentially get the source stream multiple times (while it's only possible to read meaningfully from it once).

Reply:
sure, Il remove it from the read stream